ocaml/xapi/test_network_event_loop.ml

```diff
@@ -0,0 +1,165 @@
+  received_params := None;
+  fail_firewall_update := true;
+  let network2 = Test_common.make_network ~__context ~purpose:[`nbd] ~bridge:"bridge2" () in
+  assert_not_called "network2 has no PIF connected to this host" ();
+  let _ : _ API.Ref.t = Test_common.make_pif ~__context ~network:network2 ~host:localhost ~device:"network2_pif1" () in
+  (* After [delay] seconds, the event loop should have noticed the change and
+     failed after trying to update the firewall *)
+  Thread.delay delay;
+  fail_firewall_update := false;
+  (* Test that we do wait for the given delay in case of failures *)
+  assert_not_called "The event loop should wait for the given delay in case of failures" ();
+  (* Now wait for the event loop to reregister and continue *)
+  Thread.delay network_event_loop_wait_after_failure_seconds;
+  (* Test that transient failures in the script won't stop the event loop, and
+     that we will eventually process the event we missed *)
+  assert_received_params "we should have noticed the addition of network2 and its PIF after the transient failure" ["bridge1"; "bridge2"];
+
+  (* Add a new network "network3" with a PIF connected to another host *)
+  received_params := None;
+  let network3 = Test_common.make_network ~__context ~purpose:[`nbd] ~bridge:"bridge3" () in
+  assert_not_called "network3 has no PIF connected to this host" ();
+  let _ : _ API.Ref.t = Test_common.make_pif ~__context ~network:network3 ~host:other_host ~device:"network3_pif1" () in
+  assert_not_called "pif1 of network3 is connected to another host" ();
+
+  (* Add another PIF to network3 connected to another host *)
+  received_params := None;
+  let _ : _ API.Ref.t = Test_common.make_pif ~__context ~network:network3 ~host:other_host ~device:"network3_pif2" () in
+  assert_not_called "pif2 of network3 is connected to another host" ();
+
+  (* Add a PIF to network3 connected to this host *)
+  received_params := None;
+  let pif3 = Test_common.make_pif ~__context ~network:network3 ~host:localhost ~device:"network3_pif3" () in
+  assert_received_params "network3 is now connected to this host by pif3" ["bridge1"; "bridge2"; "bridge3"];
+
+  (* Remove pif1 on network1 *)
+  received_params := None;
+  Db.PIF.destroy ~__context ~self:pif1;
+  assert_not_called "network1 still has a PIF connected to this host, so we shouldn't disallow its bridge" ();
+
+  (* Remove network1 *)
+  received_params := None;
+  Db.Network.destroy ~__context ~self:network1;
+  assert_received_params "network1 has been removed" ["bridge2"; "bridge3"];
+
+  (* Remove pif3 of network3 *)
+  received_params := None;
+  Db.PIF.destroy ~__context ~self:pif3;
+  (* This was network3's only PIF connected to this host, so we should have disabled NBD on network3's bridge *)
+  assert_received_params "network3's only PIF connected to this host has been removed" ["bridge2"];
+
+  (* Remove network2 *)
+  received_params := None;
+  fail_firewall_update := true;
```

Comment:
There's code earlier that tests the update-failure handling. Do we gain anything by doing it again here?

Reply:
Yes, the comment also says what's different here: we have an empty list of interfaces, unlike in the previous case.

Reply:
Ah, yes, the "even if it the empty list" part of the comment. Right then.